Add securityRestrictionMode API and User Default for Enhanced Security

darryl-apple opened the pull request:
#### 5d96852ac2b90a5ce465e91e268f6cb9599bfba0
<pre>
Add securityRestrictionMode API and User Default for Enhanced Security
<a href="https://bugs.webkit.org/show_bug.cgi?id=301382">https://bugs.webkit.org/show_bug.cgi?id=301382</a>
<a href="https://rdar.apple.com/163298679">rdar://163298679</a>

Reviewed by Per Arne Vollan.

Introduces public securityRestrictionMode property to replace the deprecated
_enhancedSecurityEnabled SPI. Adds WKSecurityRestrictionMode User Default
to globally enable/disable the feature for testing.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(isEnhancedSecurityFeatureEnabled):
(-[WKWebpagePreferences _setEnhancedSecurityEnabled:]):
(-[WKWebpagePreferences _enhancedSecurityEnabled]):
(-[WKWebpagePreferences setSecurityRestrictionMode:]):
(-[WKWebpagePreferences securityRestrictionMode]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
(SecurityRestrictionMode.backingSecurityRestrictionMode):
(SecurityRestrictionMode.securityRestrictionMode):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::enableEnhancedSecurityForTesting):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnablesTrue)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnableFalse)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityDisablesJIT)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterNavigation)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySamePage)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPool)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPoolReverse)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessVariantMatchesConfiguration)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessCanLaunch)):
(TestWebKitAPI::TEST(EnhancedSecurity, CaptivePortalProcessCanLaunch)):

Canonical link: <a href="https://commits.webkit.org/302121@main">https://commits.webkit.org/302121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a9a53433490586d18dccf4ce898aaf6136ecf32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128086 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38918 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135454 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/287 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/243 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/97508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/25711c33-b1d5-4124-afb5-32d3ba421ce2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131034 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114738 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78077 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78764 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108531 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137943 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/224 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/254 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111081 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105773 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26959 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/165 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29637 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52402 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/270 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/230 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

gavin-apple commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -96,6 +96,35 @@ extension WebPage {
+        /// This preference only applies to main frame navigations and will be ignored for subframe navigations. When set for a main frame, all subframe content and opened windows inherit the same security restrictions.
+        /// When the system has chosen `SecurityRestrictionMode.lockdown` (e.g., in Lockdown Mode), attempts to set a less restrictive mode will fail silently.
+        /// The default value is `SecurityRestrictionMode.none`.
+        @available(iOS 26.4, macOS 26.4, visionOS 26.4, *)
```
We should leave these TBA for now

gavin-apple commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -128,6 +157,20 @@ extension WebPage.NavigationPreferences.UpgradeToHTTPSPolicy {
     }
 }
 
+@available(iOS 26.4, macOS 26.4, visionOS 26.4, *)
```
ditto

gavin-apple commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -138,6 +181,9 @@ extension WebPage.NavigationPreferences {
 
         self.allowsContentJavaScript = wrapped.allowsContentJavaScript
         self.isLockdownModeEnabled = wrapped.isLockdownModeEnabled
+        if #available(iOS 26.4, macOS 26.4, visionOS 26.4, *) {
```
ditto

pvollan commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -828,4 +840,36 @@ - (BOOL)_allowsJSHandleCreationInPageWorld
 
+- (void)setSecurityRestrictionMode:(WKSecurityRestrictionMode)mode
+{
+    if (!isEnhancedSecurityFeatureEnabled())
+        return;
```
I am curious why we are returning early here. Shouldn't we set the mode unconditionally?

darryl-apple replied:
If we set the mode unconditionally the API could be used to change the mode. Returning early allows us to use the UserDefault to disable the API

pvollan commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -164,6 +164,11 @@ static _WKWebsiteModalContainerObservationPolicy modalContainerObservationPolicy
 
+static BOOL isEnhancedSecurityFeatureEnabled()
+{
+    return [[NSUserDefaults standardUserDefaults] boolForKey:@"WKSecurityRestrictionMode"];
```
I wonder if we should use a different name of the bool preference since `WKSecurityRestrictionMode` is defined as an enum in the API?

darryl-apple replied:
Updated to now be:
EnhancedSecurityFeatureEnabled

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -128,6 +158,20 @@ extension WebPage.NavigationPreferences.UpgradeToHTTPSPolicy {
 
+@available(WK_IOS_TBA, WK_MAC_TBA, WK_XROS_TBA, *)
+extension WebPage.NavigationPreferences.SecurityRestrictionMode {
+    init(_ wrapped: WKSecurityRestrictionMode) {
```
This should probably be `@MainActor`?

rr-codes replied:
I don’t think it’s needed since neither type needs to be main-actor-isolated

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -138,6 +182,9 @@ extension WebPage.NavigationPreferences {
 
         self.allowsContentJavaScript = wrapped.allowsContentJavaScript
         self.isLockdownModeEnabled = wrapped.isLockdownModeEnabled
+        if #available(WK_IOS_TBA , WK_MAC_TBA , WK_XROS_TBA , *) {
```
Is the availability check really needed for our own interface?

rr-codes replied:
afaik, it is always wrong to do it this way. The proper way is with compile time checks

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -138,6 +182,9 @@ extension WebPage.NavigationPreferences {
 
         self.allowsContentJavaScript = wrapped.allowsContentJavaScript
         self.isLockdownModeEnabled = wrapped.isLockdownModeEnabled
+        if #available(WK_IOS_TBA , WK_MAC_TBA , WK_XROS_TBA , *) {
```
If it is,
```suggestion
        if #available(WK_IOS_TBA, WK_MAC_TBA, WK_XROS_TBA, *) {
```

darryl-apple replied:
I added these spaces to satisfy the style checker. I agree they are not needed, shall we remove and ignore the style checker?

rr-codes replied:
see https://github.com/WebKit/WebKit/pull/52915#discussion_r2462088819 ; this is invalid and should be removed and replaced with a compile time check

pxlcoder replied:
I'm not sure if we even need a compile time check here? I don't see any obvious OS dependency.

rr-codes replied:
ah, I thought it would complain about using the property due to it’s availability annotation, but if it doesn’t, even better

darryl-apple replied:
Removed as suggested

rr-codes commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -96,6 +96,36 @@ extension WebPage {
         }
+
+        /// Security restriction modes for WebView content.
+        public enum SecurityRestrictionMode: Sendable {
```
this needs api annotations

rr-codes replied:
nested types should be before instance variables

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
```diff
@@ -32,14 +32,22 @@
+{
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"EnhancedSecurityFeatureEnabled"];
```
Once this executes, the feature will be flipped on for all subsequent tests, even those outside this file.

I think you should instead create a test subclass, use `TEST_F`, and use `SetUp` and `TearDown` methods to set/clear the default. See `WebCoreNSURLSessionTest` for an example.

darryl-apple replied:
Nice! Yep this is a good suggestion, done :)

darryl-apple committed:
Add securityRestrictionMode API and User Default for Enhanced Security
https://bugs.webkit.org/show_bug.cgi?id=301382
rdar://163298679

Reviewed by Per Arne Vollan.

Introduces public securityRestrictionMode property to replace the deprecated
_enhancedSecurityEnabled SPI. Adds WKSecurityRestrictionMode User Default
to globally enable/disable the feature for testing.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(isEnhancedSecurityFeatureEnabled):
(-[WKWebpagePreferences _setEnhancedSecurityEnabled:]):
(-[WKWebpagePreferences _enhancedSecurityEnabled]):
(-[WKWebpagePreferences setSecurityRestrictionMode:]):
(-[WKWebpagePreferences securityRestrictionMode]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
(SecurityRestrictionMode.backingSecurityRestrictionMode):
(SecurityRestrictionMode.securityRestrictionMode):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::enableEnhancedSecurityForTesting):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnablesTrue)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnableFalse)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityDisablesJIT)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterNavigation)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySamePage)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPool)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPoolReverse)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessVariantMatchesConfiguration)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessCanLaunch)):
(TestWebKitAPI::TEST(EnhancedSecurity, CaptivePortalProcessCanLaunch)):

Canonical link: https://commits.webkit.org/302121@main
